docs/posts/ibis-duckdb-geospatial/index.qmd

```diff
@@ -0,0 +1,242 @@
+
+if not db_path.exists():
+    with ZipFile(zip_path) as zip_file:
+        zip_file.extract("nyc_data.db")
+```
+
+## Let's get started
+
+The beauty of spatial databases is that they allow us to both store *and* compute over geometries.
+
+```{python}
+import ibis
+from ibis import _
+
+ibis.options.interactive = True
+
+con = ibis.duckdb.connect("nyc_data.db")
+con.list_tables()
+```
+
+We have multiple tables with information about New York City. Following Dr. Wu's class, we'll take a look at some
+spatial relations.
+
+We can start by taking a peak at the `nyc_subway_stations` table.
+
+```{python}
+subway_stations = con.table("nyc_subway_stations")
+subway_stations
+```
+
+Notice that the last column has a `geometry` type, and in this case it contains points that represent the location of
+each subway station. Let's grab the entry for the Broad St subway station.
+
+```{python}
+broad_station = subway_stations.filter(subway_stations.NAME == "Broad St")
+broad_station
+```
+
+### `geo_equals` (`ST_Equals`)
+
+In DuckDB `ST_Equals` returns `True` if two geometries are topologically equal. This means that they have the same
+dimension and identical coordinate values, although the order of the vertices may be different.
+
+The following is a bit redundant but we can check if our `"Broad St"` point matches only one point in our data using
+`geo_equals`
+
+```{python}
+subway_stations.filter(subway_stations.geom.geo_equals(broad_station.geom))
+```
+
+We can also write this query without using `broad_station` as a variable, and with the help of the deferred expressions
+API, also known as [the underscore API](https://ibis-project.org/how-to/analytics/chain_expressions.html).
+
+```{python}
+subway_stations.filter(_.geom.geo_equals(_.filter(_.NAME == "Broad St").geom))
+```
+
+### `intersect` (ST_Intersect)
+
+Let's locate the neighborhood of the "Broad Street" subway station using the
+geospatial `intersect` function. The `intersect` function returns `True` if two geometries have any points in common.
+
+```{python}
+boroughs = con.table("nyc_neighborhoods")
+boroughs
+```
+
+```{python}
+boroughs.filter(boroughs.geom.intersects(broad_station.select(broad_station.geom).to_array()))
+```
+
+### `d_within` (ST_DWithin)
+
+We can also find the streets near (say, within 10 meters) the Broad St subway station using the `d_within`
+function. The `d_within` function returns True if the geometries are within a given distance.
+
+```{python}
+streets = con.table("nyc_streets")
+streets
+```
+
+Using the deferred API, we can check which streets are within `d=10` meters of distance.
+
+```{python}
+sts_near_broad =streets.filter(_.geom.d_within(broad_station.select(_.geom).to_array(), 10))
+sts_near_broad
+```
+
+::: {.callout-note}
+In the previous query, `streets` and `broad_station` are different tables. We use [`to_array()`](https://ibis-project.org/reference/expression-tables.html#ibis.expr.types.relations.Table.to_array) to generate a
```

Comment:
```suggestion
In the previous query, `streets` and `broad_station` are different tables. We use [`to_array()`](../../reference/expression-tables.qmd#ibis.expr.types.relations.Table.to_array) to generate a
```

We should refer to the local file and let quarto worry about generating the right HTML for it.